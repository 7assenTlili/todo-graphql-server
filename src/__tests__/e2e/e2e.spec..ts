import { FetchResult, Observable, toPromise} from "apollo-link";
import initApollo from "../../infrastructure/connection/http/apollo";
import startTestServer from "./startTestServer";
import {connect, clearDatabase, closeDatabase} from "./dbUtils";
import {LOGIN_MUTATION, SIGNUP_MUTATION} from "./mutations";

beforeAll(async () => {
  await connect()
});

afterAll(async (done) => {
  await clearDatabase();
  await closeDatabase();
  done();
});

describe("server e2e", () => {
  let stop: { (): any; (): any; }, graphql: { (arg0: { query: any; variables: { pageSize: number; after: string; }; }): Observable<unknown>; ({ query, variables }: { query: any; variables?: {} | undefined; }): Observable<FetchResult<{ [key: string]: any; }, Record<string, any>, Record<string, any>>>; };

  beforeEach(async () => {
    const server = initApollo();
    const testServer = await startTestServer(server);
    stop = testServer.stop;
    graphql = testServer.graphql;
  });

  afterEach(() => stop());

  describe("User mutations", () => {
    it('should create a user document if input is valid', async () => {
      const res = await toPromise(
        graphql({
          query: SIGNUP_MUTATION,
          variables: {
            "input": {
              "email": "valid@email.com",
              "username": "hassen",
              "password": "123456789"
            }
          },
        }),
      );

      expect(res).toMatchSnapshot();
    });

    it('should not create a user document if email is invalid', async () => {
      const res = await toPromise(
        graphql({
          query: SIGNUP_MUTATION,
          variables: {
            "input": {
              "email": "invalid",
              "username": "hassen",
              "password": "123456789"
            }
          },
        }),
      );

      expect(res).toMatchSnapshot();
    });

    it('should not create a user document if username is invalid', async () => {
      const res = await toPromise(
        graphql({
          query: SIGNUP_MUTATION,
          variables: {
            "input": {
              "email": "valid@email.com",
              "username": "ha",
              "password": "123456789"
            }
          },
        }),
      );

      expect(res).toMatchSnapshot();
    });

    it('should not create a user document if password is invalid', async () => {
      const res = await toPromise(
        graphql({
          query: SIGNUP_MUTATION,
          variables: {
            "input": {
              "email": "valid@email.com",
              "username": "hassen",
              "password": "123"
            }
          },
        }),
      );

      expect(res).toMatchSnapshot();
    });

    it('should login if credentials are valid', async () => {
      const signup = await toPromise(
        graphql({
          query: SIGNUP_MUTATION,
          variables: {
            "input": {
              "email": "test@email.com",
              "username": "test",
              "password": "123456789"
            }
          },
        }),
      );

      const res = await toPromise(
        graphql({
          query: LOGIN_MUTATION,
          variables: {
            "email": "test@email.com",
            "password": "123456789"
          },
        }),
      );

      expect(res).toMatchSnapshot();
    });

    it('should not login if credentials are invalid', async () => {
      const signup = await toPromise(
        graphql({
          query: SIGNUP_MUTATION,
          variables: {
            "input": {
              "email": "test2@email.com",
              "username": "test",
              "password": "123456789"
            }
          },
        }),
      );

      const res = await toPromise(
        graphql({
          query: LOGIN_MUTATION,
          variables: {
            "email": "test2@email.com",
            "password": "123456"
          },
        }),
      );

      expect(res).toMatchSnapshot();
    });
  });
});
