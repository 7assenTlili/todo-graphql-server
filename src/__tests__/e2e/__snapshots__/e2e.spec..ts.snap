// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`server e2e User mutations should create a user document if input is valid 1`] = `
Object {
  "data": Object {
    "signup": Object {
      "email": "valid@email.com",
      "username": "hassen",
    },
  },
}
`;

exports[`server e2e User mutations should login if credentials are valid 1`] = `
Object {
  "data": Object {
    "login": Object {
      "token": "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJfaWQiOiI2MGIxNDQwMDQwNDI5ZTJmODQzOTQ0MGEiLCJlbWFpbCI6InRlc3RAZW1haWwuY29tIiwidXNlcm5hbWUiOiJ0ZXN0IiwiaWF0IjoxNjIyMjMwMDE2fQ.gfj9DByaTVQrEK5wD7Aly3ABN6V4AOc__fTCN68QlDE",
    },
  },
}
`;

exports[`server e2e User mutations should not create a user document if email is invalid 1`] = `
Object {
  "data": Object {
    "signup": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "INTERNAL_SERVER_ERROR",
      },
      "locations": Array [
        Object {
          "column": 3,
          "line": 2,
        },
      ],
      "message": "INVALID_EMAIL",
      "path": Array [
        "signup",
      ],
    },
  ],
}
`;

exports[`server e2e User mutations should not create a user document if password is invalid 1`] = `
Object {
  "data": Object {
    "signup": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "INTERNAL_SERVER_ERROR",
      },
      "locations": Array [
        Object {
          "column": 3,
          "line": 2,
        },
      ],
      "message": "INVALID_PASSWORD: should be at least 8 characters",
      "path": Array [
        "signup",
      ],
    },
  ],
}
`;

exports[`server e2e User mutations should not create a user document if username is invalid 1`] = `
Object {
  "data": Object {
    "signup": null,
  },
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "INTERNAL_SERVER_ERROR",
      },
      "locations": Array [
        Object {
          "column": 3,
          "line": 2,
        },
      ],
      "message": "INVALID_USERNAME: should be at least 3 characters",
      "path": Array [
        "signup",
      ],
    },
  ],
}
`;

exports[`server e2e User mutations should not login if credentials are invalid 1`] = `
Object {
  "data": null,
  "errors": Array [
    Object {
      "extensions": Object {
        "code": "INTERNAL_SERVER_ERROR",
      },
      "locations": Array [
        Object {
          "column": 3,
          "line": 2,
        },
      ],
      "message": "INVALID_PASSWORD: should be at least 8 characters",
      "path": Array [
        "login",
      ],
    },
  ],
}
`;
